lib/client.js

```diff
@@ -874,15 +874,22 @@ MqttClient.prototype._handlePublish = function (packet, done) {
       })
       break
     case 1:
-      // do not wait sending a puback
-      // no callback passed
-      this._sendPacket({
-        cmd: 'puback',
-        messageId: mid
-      })
-      /* falls through */
+      try {
+        // emit the message event
+        this.emit('message', topic, message, packet)
+        this.handleMessage(packet, done)
+        // send 'puback' if the above message handlers are executed
+        // successfully.
+        this._sendPacket({
+          cmd: 'puback',
+          messageId: mid
+        })
+      } catch (e) {
+        done()
+      }
```

Comment:
remove this try-catch.

We should also allocate a closure:

```js
this.emit('message', topic, message, packet)
this.handleMessage(packet, (err) => {
  if (err) {
    return done(err)
  }
  this._sendPacket({
     cmd: 'puback',
     messageId: mid
  })
})
```

Reply:
@mcollina Thanks for taking time to review the PR and the feedback. I've added the fixes proposed and submitted.